File: engine/src/elements/Border.ts
import Element from '../Element'
import Size from '../Size'
import BoxConstraints from '../BoxConstraints'
import Position from '../Position'
import Matrix from '../Matrix'

interface IRadius {
	width: number // TODO relative
	height: number
}

class Radius {
	width: number
	height: number

	constructor({ width, height }: IRadius) {
		this.width = width
		this.height = height
	}

	isZero(): boolean {
		return this.width === 0 && this.height === 0
	}

	inner(d: number): Radius {
		if (this.width <= d || this.height <= d) return Radius.zero
		return new Radius({ width: this.width - d, height: this.height - d })
	}

	static circular(radius: number): Radius {
		return new Radius({ width: radius, height: radius })
	}

	static zero = new Radius({ width: 0, height: 0 })
}

interface IBoxRadius {
	bottomLeft: Radius
	bottomRight: Radius
	topLeft: Radius
	topRight: Radius
}

class BoxRadius {
	bottomLeft: Radius
	bottomRight: Radius
	topLeft: Radius
	topRight: Radius

	constructor(radiuses: Partial<IBoxRadius>) {
		;(this.bottomLeft = radiuses.bottomLeft || Radius.zero),
			(this.bottomRight = radiuses.bottomRight || Radius.zero),
			(this.topLeft = radiuses.topLeft || Radius.zero),
			(this.topRight = radiuses.topRight || Radius.zero)
	}

	isSquare(): boolean {
		return [
			this.bottomLeft,
			this.bottomRight,
			this.topLeft,
			this.topRight
		].every(r => r.isZero())
	}

	static all(radius: Radius): BoxRadius {
		return new BoxRadius({
			bottomLeft: radius,
			bottomRight: radius,
			topLeft: radius,
			topRight: radius
		})
	}
}

interface IBorderSide {
	width: number
	color: string
}

class BorderSide {
	width: number
	color: string

	constructor(data: IBorderSide) {
		Object.assign(this, data)
	}

	static none = new BorderSide({ width: 0, color: null })

	static isEqual(a: BorderSide, b: BorderSide): boolean {
		return a.width === b.width && a.color === b.color
	}
}

interface IBoxBorder {
	top: BorderSide
	bottom: BorderSide
	left: BorderSide
	right: BorderSide
}

class BoxBorder {
	constructor({ top, bottom, left, right }: Partial<IBoxBorder>) {
		this.top = top || BorderSide.none
		this.bottom = bottom || BorderSide.none
		this.left = left || BorderSide.none
		this.right = right || BorderSide.none
	}

	top: BorderSide
	bottom: BorderSide
	left: BorderSide
	right: BorderSide

	static all({ width, color }: IBorderSide) {
		const side = new BorderSide({ width, color })
		return new BoxBorder({
			top: side,
			bottom: side,
			left: side,
			right: side
		})
	}
}

class PropsError extends Error {
	constructor(message) {
		super(message)
		this.message = message
	}
}

const degToRad = (deg: number): number => (Math.PI / 180) * deg

interface BorderProps {
	border: BoxBorder
	radius: BoxRadius
	child: Element<any>
}

const sides = ['top', 'right', 'bottom', 'left']
const corners = ['topLeft', 'topRight', 'bottomRight', 'bottomLeft']

const getIndex = (arr, i) => {
	if (i < 0) return arr[arr.length - 1]
	if (i >= arr.length) return arr[i % arr.length]
	return arr[i]
}

const checkBorders = ({ border, radius }: Partial<BorderProps>) => {
	for (let i = 0; i < 4; i++) {
		const corner = radius[getIndex(corners, i)]
		const leftSide = border[getIndex(sides, i)]
		const rightSide = border[getIndex(sides, i - 1)]
		if (!corner.isZero() && !BorderSide.isEqual(leftSide, rightSide)) {
			throw new PropsError(
				'Two adjacent borders can have rounded corner oonly if they ' +
					'have same thickness and color'
			)
		}
	}
}

class Border extends Element<BorderProps> {
	setProps(props: Partial<BorderProps>) {
		super.setProps(props)
		checkBorders(props)
	}

	layout(constraints: BoxConstraints) {
		const { border, child } = this.props
		const widths: { [name in keyof IBoxBorder]?: number } = {}
		for (let side of ['left', 'right', 'top', 'bottom']) {
			widths[side] = border[side] ? border[side].width : 0
		}
		const childConstraints = new BoxConstraints({
			minWidth: 0,
			maxWidth: constraints.maxWidth - widths.left - widths.right,
			minHeight: 0,
			maxHeight: constraints.maxHeight - widths.top - widths.bottom
		})
		const size = this.document.layoutElement(child, childConstraints)
		child.size = size
		child.relPosition = new Position({
			top: widths.top,
			left: widths.left
		})
		return new Size({
			width: size.width + widths.left + widths.right,
			height: size.height + widths.top + widths.bottom
		})
	}

	clipPath: Path2D
	transform: Matrix
	clipTransform: Matrix
	ctx: CanvasRenderingContext2D

	paint() {
		const { child, border, radius } = this.props

		const shouldClip = !radius.isSquare()
		if (shouldClip) {
			this.clipPath = new Path2D()
			this.clipTransform = new Matrix().translate(
				this.absPosition.left,
				this.absPosition.top
			)
			this.iterateSides(this.clipSide.bind(this))
		}

		this.ctx = this.document.getLayer().ctx
		this.ctx.save()
		this.document.paintElement(this.props.child, {
			clip: shouldClip && this.clipPath
		})

		this.transform = new Matrix()
		this.ctx.translate(this.absPosition.left, this.absPosition.top)
		this.iterateSides(this.paintSide.bind(this))
		this.ctx.restore()
	}

	iterateSides(callback) {
		const { border, radius } = this.props
		for (let i = 0; i < 4; i++) {
			callback({
				side: border[getIndex(sides, i)],
				prevSide: border[getIndex(sides, i - 1)],
				nextSide: border[getIndex(sides, i + 1)],
				leftRadius: radius[getIndex(corners, i)],
				rightRadius: radius[getIndex(corners, i + 1)],
				rotation: i * 90
			})
		}
	}

	clipSide({ prevSide, side, nextSide, leftRadius, rightRadius, rotation }) {
		const width = rotation % 180 === 0 ? this.size.width : this.size.height
		const begin = Math.max(prevSide.width, leftRadius.width)
		const end = width - Math.max(nextSide.width, rightRadius.width)
		this.clipPath.lineTo(...this.clipTransform.calc(begin, side.width))
		this.clipPath.lineTo(...this.clipTransform.calc(end, side.width))
		const innerRadius = rightRadius.inner(side.width)
		if (!innerRadius.isZero()) {
			// @ts-ignore error about number of arguments
			this.clipPath.ellipse(
				...this.clipTransform.calc(
					width - rightRadius.width,
					rightRadius.height
				),
				// making radius 1px bigger is needed to avoid artifacts 
				// between border and child
				innerRadius.width + 1,
				innerRadius.height + 1,
				0,
				degToRad(rotation - 90), // start angle
				degToRad(rotation) // end angle
			)
		}
		this.clipTransform.translate(width, 0).rotate(degToRad(90), 0, 0)
	}

	paintArc({ radius, side, width, rotation }) {
		this.ctx.beginPath()
		// todo radius width/height will change
		const lineWidth = Math.min(side.width, radius.width) // height???
		this.ctx.lineWidth = lineWidth
		// @ts-ignore error about number of arguments
		this.ctx.ellipse(
			...this.transform.calc(width - radius.width, radius.height), // left, top
			radius.width - lineWidth / 2, // radius x
			radius.height - lineWidth / 2, // radius y
			0, // ellipse rotation
			degToRad(rotation - 90), // start angle
			degToRad(rotation) // end angle
		)
		this.ctx.stroke()
	}

	paintTriangle({ radius, side, prevSide, nextSide, width, edge }) {
		if (edge === 'left' && prevSide.width === 0) return
		if (edge === 'right' && nextSide.width === 0) return
		this.ctx.beginPath()
		if (edge === 'left') {
			this.ctx.moveTo(...this.transform.calc(0, 0))
			this.ctx.lineTo(...this.transform.calc(prevSide.width, 0))
			this.ctx.lineTo(...this.transform.calc(prevSide.width, side.width))
		} else {
			this.ctx.moveTo(...this.transform.calc(width, 0))
			this.ctx.lineTo(...this.transform.calc(width - nextSide.width, 0))
			this.ctx.lineTo(
				...this.transform.calc(width - nextSide.width, side.width)
			)
		}
		this.ctx.fill()
	}

	paintSide({ prevSide, side, nextSide, leftRadius, rightRadius, rotation }) {
		const width = rotation % 180 === 0 ? this.size.width : this.size.height
		if (side.width === 0) {
			this.transform.translate(width, 0).rotate(degToRad(90), 0, 0)
			return
		}

		this.ctx.strokeStyle = side.color
		this.ctx.fillStyle = side.color
		if (leftRadius.isZero() && prevSide.color !== side.color) {
			this.paintTriangle({
				prevSide,
				side,
				nextSide,
				width,
				edge: 'left',
				radius: leftRadius
			})
		}
		this.ctx.lineWidth = side.width
		this.ctx.beginPath()
		this.ctx.moveTo(
			...this.transform.calc(
				leftRadius.isZero()
					? prevSide.color === side.color
						? 0
						: prevSide.width
					: leftRadius.width,
				side.width / 2
			)
		)
		this.ctx.lineTo(
			...this.transform.calc(
				width -
					(rightRadius.isZero() ? nextSide.width : rightRadius.width),
				side.width / 2
			)
		)
		this.ctx.stroke()
		if (rightRadius.isZero()) {
			if (prevSide.color !== side.color) {
				this.paintTriangle({
					prevSide,
					side,
					nextSide,
					width,
					edge: 'right',
					radius: rightRadius
				})
			}
		} else {
			this.paintArc({ side, width, radius: rightRadius, rotation })
		}
		this.transform.translate(width, 0).rotate(degToRad(90), 0, 0)
	}
}

export default Border
export { Radius, BoxRadius, BorderSide, BoxBorder }

File: core/src/elements/border.cpp
#include "border.hpp"

namespace aardvark::elements {

Border::Border(std::shared_ptr<Element> child, BoxBorders borders,
               BoxRadiuses radiuses, bool is_repaint_boundary)
    : SingleChildElement(child, is_repaint_boundary,
                         /* size_depends_on_parent */ false),
      borders(borders),
      radiuses(radiuses){};

Size Border::layout(BoxConstraints constraints) {
    auto vert_width = borders.left.width + borders.right.width;
    auto horiz_width = borders.top.width + borders.bottom.width;
    auto child_constraints = BoxConstraints{
        0,                                    // min_width
        constraints.max_width - vert_width,   // max_width
        0,                                    // min_height
        constraints.max_height - horiz_width  // max_height
    };
    auto size = document->layout_element(child.get(), child_constraints);
    child->size = size;
    child->rel_position = Position{
        static_cast<float>(borders.left.width),  // left
        static_cast<float>(borders.top.width)    // top
    };
    return Size{
        size.width + vert_width,   // width
        size.height + horiz_width  // height
    };
};

void Border::paint(bool is_changed) {
    auto layer = document->get_layer();
    document->setup_layer(layer, this);
    this->canvas = layer->canvas;

    // After painting each border side, coordinates are translated and
    // rotated 90 degrees, so next border can be painted with same function.
    matrix = SkMatrix();
    rotation = 0;
    // top -> right -> bottom -> left
    paint_side(borders.left, borders.top, borders.right, radiuses.topLeft,
               radiuses.topRight);
    paint_side(borders.top, borders.right, borders.bottom, radiuses.topRight,
               radiuses.bottomRight);
    paint_side(borders.right, borders.bottom, borders.left,
               radiuses.bottomRight, radiuses.bottomLeft);
    paint_side(borders.bottom, borders.left, borders.top, radiuses.bottomLeft,
               radiuses.topLeft);

    // TODO if all *inner* radiuses are square
    bool need_custom_clip = !radiuses.is_square();
    if (need_custom_clip) {
        clip_matrix = SkMatrix();
        clip_path = SkPath();
        rotation = 0;
        // top -> right -> bottom -> left
        clip_side(borders.left, borders.top, borders.right, radiuses.topLeft,
                  radiuses.topRight);
        clip_side(borders.top, borders.right, borders.bottom, radiuses.topRight,
                  radiuses.bottomRight);
        clip_side(borders.right, borders.bottom, borders.left,
                  radiuses.bottomRight, radiuses.bottomLeft);
        clip_side(borders.bottom, borders.left, borders.top,
                  radiuses.bottomLeft, radiuses.topLeft);
    }
    child->clip = need_custom_clip ? std::optional(clip_path) : std::nullopt;
    document->paint_element(child.get());
};

SkPoint calc(SkMatrix& matrix, float left, float top) {
    SkPoint point{left, top};
    matrix.mapPoints(&point, 1);
    return point;
};

void transform_to_next_side(SkMatrix& matrix, float side_width) {
    // For some unknown reason `preTranslate` is not working same as `preConcat`
    SkMatrix m;
    m.setTranslate(side_width, 0);
    matrix.preConcat(m);
    matrix.preRotate(90);
};

void Border::paint_triangle(Radius& radius, BorderSide& prev_side,
                            BorderSide& side, BorderSide& next_side, int width,
                            bool is_left_edge) {
    if (is_left_edge && prev_side.width == 0) return;
    if (!is_left_edge && next_side.width == 0) return;

    SkPath path;
    if (is_left_edge) {
        path.moveTo(calc(matrix, 0, 0));
        path.lineTo(calc(matrix, prev_side.width, 0));
        path.lineTo(calc(matrix, prev_side.width, side.width));
    } else {
        path.moveTo(calc(matrix, width, 0));
        path.lineTo(calc(matrix, width - next_side.width, 0));
        path.lineTo(calc(matrix, width - next_side.width, side.width));
    }

    SkPaint paint;
    paint.setStyle(SkPaint::kFill_Style);
    paint.setColor(side.color);
    paint.setAntiAlias(true);
    canvas->drawPath(path, paint);
};

void Border::paint_arc(Radius& radius, BorderSide& side, int width) {
    // TODO change radius width/height to opposite when rotated
    auto line_width = fmin(side.width, radius.width);  // height???
    SkPaint arc_paint;
    arc_paint.setColor(side.color);
    arc_paint.setStyle(SkPaint::kStroke_Style);
    arc_paint.setStrokeWidth(line_width);
    arc_paint.setAntiAlias(true);
    // Adding line_width/2 is needed to paint stroke inside of the oval
    auto bounds = SkRect::MakeLTRB(width - 2 * radius.width,  // l
                                   line_width / 2,            // t
                                   width - line_width / 2,    // r
                                   2 * radius.height          // b
    );
    matrix.mapRect(&bounds);
    canvas->drawArc(bounds,         // oval
                    rotation - 90,  // startAngle, 0 is right middle
                    90,             // sweepAngle
                    false,          // useCenter
                    arc_paint       // paint
    );
};

void Border::paint_side(BorderSide& prev_side, BorderSide& side,
                        BorderSide& next_side, Radius& left_radius,
                        Radius& right_radius) {
    auto width = rotation % 180 == 0 ? size.width : size.height;
    if (side.width == 0) {
        transform_to_next_side(matrix, width);
        rotation += 90;
        return;
    }

    float start;
    if (left_radius.is_square()) {
        if (prev_side.color == side.color) {
            // When prev side has the same color, draw line from the beginning
            start = 0;
        } else {
            // When prev side has different color, draw triangle on the left and
            // then draw the line
            paint_triangle(left_radius, prev_side, side, next_side, width,
                           true);
            start = prev_side.width;
        }
    } else {
        start = left_radius.width;
    }
    float end = width - (right_radius.is_square() ? next_side.width
                                                  : right_radius.width);

    // Draw the line
    SkPaint line_paint;
    line_paint.setStyle(SkPaint::kStroke_Style);
    line_paint.setColor(side.color);
    line_paint.setStrokeWidth(side.width);
    line_paint.setAntiAlias(true);
    canvas->drawLine(calc(matrix, start, side.width / 2),
                     calc(matrix, end, side.width / 2), line_paint);

    // Draw different type of transition to next side depending on corner
    if (right_radius.is_square()) {
        if (side.color != next_side.color) {
            // Draw triangle on the end of the line when next side has different
            // color
            paint_triangle(right_radius, prev_side, side, next_side, width,
                           false);
        }
    } else {
        // When corner is rounded draw an arc
        paint_arc(right_radius, side, width);
    }

    transform_to_next_side(matrix, width);
    rotation += 90;
};

void Border::clip_side(BorderSide& prev_side, BorderSide& side,
                       BorderSide& next_side, Radius& left_radius,
                       Radius& right_radius) {
    auto width = (rotation % 180 == 0 ? size.width : size.height) -
                 prev_side.width - next_side.width;
    auto begin = fmax(0, left_radius.width - prev_side.width);
    auto end = width - fmax(0, right_radius.width - next_side.width);
    clip_path.lineTo(calc(clip_matrix, begin, 0));
    clip_path.lineTo(calc(clip_matrix, end, 0));
    auto inner_radius = right_radius.inner(side.width);
    if (!inner_radius.is_square()) {
        auto bounds = SkRect::MakeLTRB(width - 2 * inner_radius.width,  // left
                                       0,                               // top
                                       width,                           // right
                                       2 * inner_radius.height  // height
        );
        clip_matrix.mapRect(&bounds);
        clip_path.arcTo(bounds,         // oval bounds
                        rotation - 90,  // startAngle, 0 is right middle
                        90,             // sweepAngle
                        false           // forceMoveTo
        );
    }

    transform_to_next_side(clip_matrix, width);
    rotation += 90;
};

};  // namespace aardvark::elements
